File: apps/schoolflow/backend/app/schemas/parent_portal.py
"""Pydantic schemas for Parent Portal."""
from __future__ import annotations

from datetime import datetime, date
from typing import Any

from pydantic import BaseModel, EmailStr, Field


# ──────────────── Authentication ────────────────
class ParentRegisterRequest(BaseModel):
    """Registration request for parent account."""
    first_name: str = Field(min_length=1, max_length=100)
    last_name: str = Field(min_length=1, max_length=100)
    email: EmailStr
    phone: str | None = Field(None, max_length=30)
    password: str = Field(min_length=6)
    
    # Link to student(s) - optional at registration
    student_ids: list[int] | None = None


class ParentMagicLinkRequest(BaseModel):
    """Request magic link authentication."""
    email: EmailStr


class ParentVerifyMagicRequest(BaseModel):
    """Verify magic link token."""
    token: str


class ParentLoginRequest(BaseModel):
    """Password login request."""
    email: EmailStr
    password: str = Field(min_length=6)


class ParentTokenResponse(BaseModel):
    """Token response for parent authentication."""
    access_token: str
    refresh_token: str
    token_type: str = "bearer"
    parent: "ParentAccountOut"


class ParentRefreshRequest(BaseModel):
    """Refresh token request."""
    refresh_token: str


# ──────────────── Profile ────────────────
class ParentProfileUpdate(BaseModel):
    """Update parent profile."""
    first_name: str | None = Field(None, max_length=100)
    last_name: str | None = Field(None, max_length=100)
    phone: str | None = Field(None, max_length=30)
    address: str | None = None


class ParentAccountUpdate(BaseModel):
    """Update parent account settings."""
    preferred_language: str | None = Field(None, pattern="^(fr|en|sw|wo)$")
    notification_channels: list[str] | None = None
    receive_sms_notifications: bool | None = None
    receive_email_notifications: bool | None = None


class ParentPasswordChange(BaseModel):
    """Change password request."""
    current_password: str
    new_password: str = Field(min_length=6)


class ParentAccountOut(BaseModel):
    """Parent account output."""
    model_config = {"from_attributes": True}
    
    id: int
    email: str
    phone: str | None
    preferred_language: str
    notification_channels: list[str]
    receive_sms_notifications: bool
    receive_email_notifications: bool
    is_active: bool
    email_verified: bool
    last_login: datetime | None
    created_at: datetime
    
    # Parent info
    parent: "ParentInfoOut"


class ParentInfoOut(BaseModel):
    """Basic parent information."""
    model_config = {"from_attributes": True}
    
    id: int
    first_name: str
    last_name: str
    email: str | None
    phone: str | None
    address: str | None
    full_name: str


# ──────────────── Children ────────────────
class ChildOut(BaseModel):
    """Child (student) information for parent."""
    model_config = {"from_attributes": True}
    
    id: int
    student_number: str
    first_name: str
    last_name: str
    full_name: str
    birth_date: datetime | None
    gender: str | None
    is_active: bool
    
    # Class info
    class_id: int | None
    class_name: str | None
    class_level: str | None
    
    # Parent relationship info
    relationship: str | None
    is_primary: bool


class ChildDetailOut(ChildOut):
    """Detailed child information."""
    
    # Academic summary
    average_grade: float | None
    attendance_rate: float | None
    pending_fees: float
    total_fees: float


# ──────────────── Grades ────────────────
class GradeForParentOut(BaseModel):
    """Grade information for parent view."""
    model_config = {"from_attributes": True}
    
    id: int
    subject_id: int
    subject_name: str
    subject_coefficient: float
    term_id: int
    term_name: str
    academic_year: str
    score: float
    max_score: float
    percentage: float
    comment: str | None
    created_at: datetime


class SubjectGradesOut(BaseModel):
    """Grades grouped by subject."""
    subject_id: int
    subject_name: str
    coefficient: float
    teacher_name: str | None
    grades: list[GradeForParentOut]
    average: float


class TermGradesOut(BaseModel):
    """Grades for a term."""
    term_id: int
    term_name: str
    academic_year: str
    subjects: list[SubjectGradesOut]
    overall_average: float
    class_average: float | None
    rank: int | None


class ReportCardOut(BaseModel):
    """Report card for a term."""
    student_id: int
    student_name: str
    student_number: str
    class_name: str
    class_level: str
    
    term_id: int
    term_name: str
    academic_year: str
    
    subjects: list[SubjectGradesOut]
    overall_average: float
    class_average: float | None
    rank: int | None
    total_students: int | None
    
    # Attendance summary
    total_days: int
    present_days: int
    absent_days: int
    late_days: int
    attendance_rate: float
    
    # Comments
    teacher_comment: str | None
    principal_comment: str | None
    
    generated_at: datetime


# ──────────────── Attendance ────────────────
class AttendanceRecordForParentOut(BaseModel):
    """Attendance record for parent view."""
    model_config = {"from_attributes": True}
    
    id: int
    date: date
    status: str  # present, absent, late, excused
    arrival_time: str | None
    notes: str | None
    class_name: str


class AttendanceStatsForParentOut(BaseModel):
    """Attendance statistics for parent."""
    student_id: int
    period_start: date
    period_end: date
    
    total_days: int
    present_days: int
    absent_days: int
    late_days: int
    excused_days: int
    
    attendance_rate: float
    punctuality_rate: float
    consecutive_absences: int


class AttendanceCalendarDay(BaseModel):
    """Single day attendance status."""
    date: date
    status: str  # present, absent, late, excused, no_school
    is_weekend: bool
    is_holiday: bool


# ──────────────── Timetable ────────────────
class TimetableEntryForParentOut(BaseModel):
    """Timetable entry for parent view."""
    id: int
    day_of_week: int  # 0=Monday, 5=Saturday
    day_name: str
    time_slot_id: int
    slot_name: str
    start_time: str
    end_time: str
    subject_name: str
    teacher_name: str | None
    room: str | None


class DayTimetableOut(BaseModel):
    """Timetable for a single day."""
    day_of_week: int
    day_name: str
    entries: list[TimetableEntryForParentOut]


class WeekTimetableOut(BaseModel):
    """Full week timetable."""
    student_id: int
    student_name: str
    class_name: str
    days: list[DayTimetableOut]


# ──────────────── Fees ────────────────
class FeeInvoiceForParentOut(BaseModel):
    """Fee invoice for parent view."""
    model_config = {"from_attributes": True}
    
    id: int
    invoice_number: str
    student_id: int
    student_name: str
    term_id: int | None
    term_name: str | None
    amount: float
    description: str
    due_date: datetime | None
    status: str
    created_at: datetime
    
    # Payment info
    amount_paid: float
    amount_remaining: float
    payments: list["FeePaymentForParentOut"]


class FeePaymentForParentOut(BaseModel):
    """Fee payment for parent view."""
    model_config = {"from_attributes": True}
    
    id: int
    amount: float
    method: str
    reference: str | None
    paid_at: datetime


class FeePaymentRequest(BaseModel):
    """Payment request."""
    fee_invoice_id: int
    amount: float = Field(gt=0)
    method: str = Field(pattern="^(CASH|MOBILE_MONEY|BANK_TRANSFER|CARD)$")
    phone_number: str | None = None  # For mobile money
    provider: str | None = None  # Orange, MTN, Wave, etc.


class MobileMoneyPaymentRequest(BaseModel):
    """Mobile money payment request."""
    fee_invoice_id: int
    amount: float = Field(gt=0)
    provider: str  # orange, mtn, wave, moov, mpesa
    phone_number: str = Field(min_length=8, max_length=15)


# ──────────────── Messages ────────────────
class ParentMessageOut(BaseModel):
    """Message for parent view."""
    model_config = {"from_attributes": True}
    
    id: int
    subject: str
    content: str
    sender_type: str
    sender_name: str | None
    thread_id: str | None
    reply_to_id: int | None
    student_id: int | None
    student_name: str | None
    priority: str
    category: str | None
    read: bool
    read_at: datetime | None
    created_at: datetime


class ParentMessageCreate(BaseModel):
    """Create new message from parent."""
    subject: str = Field(min_length=1, max_length=255)
    content: str = Field(min_length=1)
    student_id: int | None = None  # Optional: related to a specific child
    category: str | None = None
    priority: str = Field(default="normal", pattern="^(low|normal|high|urgent)$")


class ParentMessageReply(BaseModel):
    """Reply to a message."""
    content: str = Field(min_length=1)


class MessageThreadOut(BaseModel):
    """Message thread."""
    thread_id: str
    subject: str
    messages: list[ParentMessageOut]
    latest_message: datetime
    unread_count: int


# ──────────────── Notifications ────────────────
class ParentNotificationOut(BaseModel):
    """Notification for parent."""
    model_config = {"from_attributes": True}
    
    id: int
    type: str
    title: str
    content: str
    student_id: int | None
    student_name: str | None
    reference_type: str | None
    reference_id: int | None
    read: bool
    read_at: datetime | None
    created_at: datetime


class NotificationListOut(BaseModel):
    """Paginated notification list."""
    items: list[ParentNotificationOut]
    total: int
    unread_count: int


# ──────────────── Dashboard ────────────────
class ParentDashboardOut(BaseModel):
    """Parent dashboard data."""
    parent: ParentInfoOut
    
    # Children count
    children_count: int
    
    # Quick stats
    unread_messages: int
    unread_notifications: int
    pending_fees_total: float
    upcoming_payments: list[FeeInvoiceForParentOut]
    
    # Recent activity
    recent_grades: list[GradeForParentOut]
    recent_attendance: list[AttendanceRecordForParentOut]
    recent_notifications: list[ParentNotificationOut]
    
    # Children summaries
    children: list[ChildDetailOut]


# ──────────────── Link Student ────────────────
class LinkStudentRequest(BaseModel):
    """Request to link a student to parent account."""
    student_number: str
    relationship: str = Field(default="parent", pattern="^(parent|guardian|other)$")


class LinkStudentVerify(BaseModel):
    """Verify student for linking (returns info without linking)."""
    student_number: str
    birth_date: date | None = None  # For verification


# ──────────────── Update forward references ────────────────
ParentAccountOut.model_rebuild()
FeeInvoiceForParentOut.model_rebuild()

File: apps/schoolflow/frontend/src/store/parentPortal.ts
import { create } from 'zustand'
import portalApi from '../lib/portalApi'

// Types
interface Parent {
  id: number
  first_name: string
  last_name: string
  email: string | null
  phone: string | null
  address: string | null
  full_name: string
}

interface ParentAccount {
  id: number
  email: string
  phone: string | null
  preferred_language: string
  notification_channels: string[]
  receive_sms_notifications: boolean
  receive_email_notifications: boolean
  is_active: boolean
  email_verified: boolean
  last_login: string | null
  created_at: string
  parent: Parent
}

interface Child {
  id: number
  student_number: string
  first_name: string
  last_name: string
  full_name: string
  birth_date: string | null
  gender: string | null
  is_active: boolean
  class_id: number | null
  class_name: string | null
  class_level: string | null
  relationship: string | null
  is_primary: boolean
}

interface ChildDetail extends Child {
  average_grade: number | null
  attendance_rate: number | null
  pending_fees: number
  total_fees: number
}

interface Grade {
  id: number
  subject_id: number
  subject_name: string
  subject_coefficient: number
  term_id: number
  term_name: string
  academic_year: string
  score: number
  max_score: number
  percentage: number
  comment: string | null
  created_at: string
}

interface AttendanceRecord {
  id: number
  date: string
  status: string
  arrival_time: string | null
  notes: string | null
  class_name: string
}

interface AttendanceStats {
  student_id: number
  total_days: number
  present_days: number
  absent_days: number
  late_days: number
  excused_days: number
  attendance_rate: number
  punctuality_rate: number
  consecutive_absences?: number
}

interface FeeInvoice {
  id: number
  invoice_number: string
  student_id: number
  student_name: string
  term_id: number | null
  term_name: string | null
  amount: number
  description: string
  due_date: string | null
  status: string
  created_at: string
  amount_paid: number
  amount_remaining: number
  payments: Array<{
    id: number
    amount: number
    method: string
    reference: string | null
    paid_at: string
  }>
}

interface Message {
  id: number
  subject: string
  content: string
  sender_type: string
  sender_name: string | null
  thread_id: string | null
  reply_to_id: number | null
  student_id: number | null
  priority: string
  category: string | null
  read: boolean
  read_at: string | null
  created_at: string
}

interface Notification {
  id: number
  type: string
  title: string
  content: string
  student_id: number | null
  student_name: string | null
  reference_type: string | null
  reference_id: number | null
  read: boolean
  read_at: string | null
  created_at: string
}

interface Dashboard {
  parent: Parent
  children_count: number
  unread_messages: number
  unread_notifications: number
  pending_fees_total: number
  upcoming_payments: FeeInvoice[]
  recent_grades: Grade[]
  recent_attendance: AttendanceRecord[]
  recent_notifications: Notification[]
  children: ChildDetail[]
}

interface TimetableEntry {
  id: number
  day_of_week: number
  day_name: string
  time_slot_id: number
  slot_name: string
  start_time: string
  end_time: string
  subject_name: string
  teacher_name: string | null
  room: string | null
}

interface DayTimetable {
  day_of_week: number
  day_name: string
  entries: TimetableEntry[]
}

interface WeekTimetable {
  student_id: number
  student_name: string
  class_name: string
  days: DayTimetable[]
}

// Store state
interface ParentPortalState {
  // Auth
  account: ParentAccount | null
  isAuthenticated: boolean
  loading: boolean

  // Children
  children: Child[]
  selectedChild: ChildDetail | null

  // Dashboard
  dashboard: Dashboard | null

  // Grades
  grades: Grade[]

  // Attendance
  attendanceRecords: AttendanceRecord[]
  attendanceStats: AttendanceStats | null

  // Timetable
  timetable: WeekTimetable | null

  // Fees
  feeInvoices: FeeInvoice[]

  // Messages
  messages: Message[]
  messagesTotal: number

  // Notifications
  notifications: Notification[]
  notificationsTotal: number
  notificationsUnread: number

  // Actions - Auth
  login: (email: string, password: string) => Promise<void>
  requestMagicLink: (email: string) => Promise<void>
  verifyMagicLink: (token: string) => Promise<void>
  logout: () => void
  initAuth: () => void

  // Actions - Profile
  updateProfile: (data: Partial<Parent>) => Promise<void>
  changePassword: (currentPassword: string, newPassword: string) => Promise<void>

  // Actions - Children
  fetchChildren: () => Promise<void>
  fetchChildDetail: (studentId: number) => Promise<void>
  linkStudent: (studentNumber: string, relationship: string) => Promise<void>

  // Actions - Dashboard
  fetchDashboard: () => Promise<void>

  // Actions - Grades
  fetchGrades: (studentId: number, termId?: number) => Promise<void>

  // Actions - Attendance
  fetchAttendance: (studentId: number, startDate?: string, endDate?: string) => Promise<void>
  fetchAttendanceStats: (studentId: number, startDate?: string, endDate?: string) => Promise<void>

  // Actions - Timetable
  fetchTimetable: (studentId: number) => Promise<void>

  // Actions - Fees
  fetchFees: (studentId: number) => Promise<void>
  payFee: (feeId: number, amount: number, method: string) => Promise<void>
  payFeeMobileMoney: (feeId: number, amount: number, provider: string, phoneNumber: string) => Promise<any>

  // Actions - Messages
  fetchMessages: (unreadOnly?: boolean, page?: number) => Promise<void>
  sendMessage: (subject: string, content: string, studentId?: number, category?: string, priority?: string) => Promise<void>
  replyToMessage: (messageId: number, content: string) => Promise<void>
  markMessageRead: (messageId: number) => Promise<void>

  // Actions - Notifications
  fetchNotifications: (unreadOnly?: boolean, page?: number) => Promise<void>
  markNotificationRead: (notificationId: number) => Promise<void>
  markAllNotificationsRead: () => Promise<void>
}

// Safe JSON parse
function safeJsonParse<T>(str: string | null): T | null {
  if (!str || str === 'undefined' || str === 'null') return null
  try {
    return JSON.parse(str)
  } catch {
    return null
  }
}

export const useParentPortalStore = create<ParentPortalState>((set, get) => ({
  // Initial state
  account: null,
  isAuthenticated: false,
  loading: false,
  children: [],
  selectedChild: null,
  dashboard: null,
  grades: [],
  attendanceRecords: [],
  attendanceStats: null,
  timetable: null,
  feeInvoices: [],
  messages: [],
  messagesTotal: 0,
  notifications: [],
  notificationsTotal: 0,
  notificationsUnread: 0,

  // Auth actions
  initAuth: () => {
    const raw = localStorage.getItem('parent_user')
    const account = safeJsonParse<ParentAccount>(raw)
    if (account) {
      set({ account, isAuthenticated: true })
    }
  },

  login: async (email, password) => {
    set({ loading: true })
    try {
      const { data } = await portalApi.post('/auth/login', { email, password })
      localStorage.setItem('parent_access_token', data.access_token)
      localStorage.setItem('parent_refresh_token', data.refresh_token)
      localStorage.setItem('parent_user', JSON.stringify(data.parent))
      set({ account: data.parent, isAuthenticated: true, loading: false })
    } catch (error) {
      set({ loading: false })
      throw error
    }
  },

  requestMagicLink: async (email) => {
    await portalApi.post('/auth/magic-link', { email })
  },

  verifyMagicLink: async (token) => {
    set({ loading: true })
    try {
      const { data } = await portalApi.post('/auth/verify-magic', { token })
      localStorage.setItem('parent_access_token', data.access_token)
      localStorage.setItem('parent_refresh_token', data.refresh_token)
      localStorage.setItem('parent_user', JSON.stringify(data.parent))
      set({ account: data.parent, isAuthenticated: true, loading: false })
    } catch (error) {
      set({ loading: false })
      throw error
    }
  },

  logout: () => {
    localStorage.removeItem('parent_access_token')
    localStorage.removeItem('parent_refresh_token')
    localStorage.removeItem('parent_user')
    set({
      account: null,
      isAuthenticated: false,
      children: [],
      selectedChild: null,
      dashboard: null,
      grades: [],
      attendanceRecords: [],
      attendanceStats: null,
      timetable: null,
      feeInvoices: [],
      messages: [],
      notifications: []
    })
  },

  // Profile actions
  updateProfile: async (data) => {
    const { data: updated } = await portalApi.put('/profile', data)
    const account = get().account
    if (account) {
      const newAccount = { ...account, parent: updated }
      localStorage.setItem('parent_user', JSON.stringify(newAccount))
      set({ account: newAccount })
    }
  },

  changePassword: async (currentPassword, newPassword) => {
    await portalApi.put('/password', {
      current_password: currentPassword,
      new_password: newPassword
    })
  },

  // Children actions
  fetchChildren: async () => {
    const { data } = await portalApi.get('/children')
    set({ children: data })
  },

  fetchChildDetail: async (studentId) => {
    const { data } = await portalApi.get(`/children/${studentId}`)
    set({ selectedChild: data })
  },

  linkStudent: async (studentNumber, relationship) => {
    await portalApi.post('/children/link', {
      student_number: studentNumber,
      relationship
    })
    // Refresh children list
    await get().fetchChildren()
  },

  // Dashboard actions
  fetchDashboard: async () => {
    const { data } = await portalApi.get('/dashboard')
    set({ dashboard: data, children: data.children })
  },

  // Grades actions
  fetchGrades: async (studentId, termId) => {
    const url = termId
      ? `/children/${studentId}/grades?term_id=${termId}`
      : `/children/${studentId}/grades`
    const { data } = await portalApi.get(url)
    set({ grades: data })
  },

  // Attendance actions
  fetchAttendance: async (studentId, startDate, endDate) => {
    const params = new URLSearchParams()
    if (startDate) params.append('start_date', startDate)
    if (endDate) params.append('end_date', endDate)
    const { data } = await portalApi.get(`/children/${studentId}/attendance?${params}`)
    set({ attendanceRecords: data })
  },

  fetchAttendanceStats: async (studentId, startDate, endDate) => {
    const params = new URLSearchParams()
    if (startDate) params.append('start_date', startDate)
    if (endDate) params.append('end_date', endDate)
    const { data } = await portalApi.get(`/children/${studentId}/attendance/stats?${params}`)
    set({ attendanceStats: data })
  },

  // Timetable actions
  fetchTimetable: async (studentId) => {
    const { data } = await portalApi.get(`/children/${studentId}/timetable`)
    set({ timetable: data })
  },

  // Fees actions
  fetchFees: async (studentId) => {
    const { data } = await portalApi.get(`/children/${studentId}/fees`)
    set({ feeInvoices: data })
  },

  payFee: async (feeId, amount, method) => {
    await portalApi.post(`/fees/${feeId}/pay`, { amount, method })
  },

  payFeeMobileMoney: async (feeId, amount, provider, phoneNumber) => {
    const { data } = await portalApi.post(`/fees/${feeId}/pay/mobile-money`, {
      amount,
      provider,
      phone_number: phoneNumber
    })
    return data
  },

  // Messages actions
  fetchMessages: async (unreadOnly = false, page = 1) => {
    const { data } = await portalApi.get(`/messages?unread_only=${unreadOnly}&page=${page}`)
    set({ messages: data.items, messagesTotal: data.total })
  },

  sendMessage: async (subject, content, studentId, category, priority = 'normal') => {
    await portalApi.post('/messages', {
      subject,
      content,
      student_id: studentId,
      category,
      priority
    })
  },

  replyToMessage: async (messageId, content) => {
    await portalApi.post(`/messages/${messageId}/reply`, { content })
  },

  markMessageRead: async (messageId) => {
    await portalApi.put(`/messages/${messageId}/read`)
    const messages = get().messages.map(m =>
      m.id === messageId ? { ...m, read: true } : m
    )
    set({ messages })
  },

  // Notifications actions
  fetchNotifications: async (unreadOnly = false, page = 1) => {
    const { data } = await portalApi.get(`/notifications?unread_only=${unreadOnly}&page=${page}`)
    set({
      notifications: data.items,
      notificationsTotal: data.total,
      notificationsUnread: data.unread_count
    })
  },

  markNotificationRead: async (notificationId) => {
    await portalApi.put(`/notifications/${notificationId}/read`)
    const notifications = get().notifications.map(n =>
      n.id === notificationId ? { ...n, read: true } : n
    )
    const unread = get().notificationsUnread - 1
    set({ notifications, notificationsUnread: Math.max(0, unread) })
  },

  markAllNotificationsRead: async () => {
    await portalApi.put('/notifications/read-all')
    const notifications = get().notifications.map(n => ({ ...n, read: true }))
    set({ notifications, notificationsUnread: 0 })
  }
}))
